pkg/controller/database/rdsinstance.go

```diff
@@ -83,56 +76,30 @@ type connector struct {
 }
 
 func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) { //nolint:gocyclo
-	// TODO(negz): This connect method will be simpler once we no longer have to
-	// account for the deprecated Provider type.
 	cr, ok := mg.(*v1alpha1.RDSInstance)
 	if !ok {
 		return nil, errors.New(errNotRDSInstance)
 	}
 
-	// TODO(negz): This connection logic should be generalised once this
-	// provider has more than one kind of managed resource.
-	var (
-		sel    *xpv1.SecretKeySelector
-		region string
-	)
-	switch {
-	case cr.GetProviderConfigReference() != nil:
-		if err := c.usage.Track(ctx, mg); err != nil {
-			return nil, errors.Wrap(err, errTrackUsage)
-		}
-
-		pc := &aliv1alpha1.ProviderConfig{}
-		if err := c.client.Get(ctx, types.NamespacedName{Name: cr.Spec.ProviderConfigReference.Name}, pc); err != nil {
-			return nil, errors.Wrap(err, errGetProviderConfig)
-		}
-		if s := pc.Spec.Credentials.Source; s != xpv1.CredentialsSourceSecret {
-			return nil, errors.Errorf(errFmtUnsupportedCredSource, s)
-		}
-		sel = pc.Spec.Credentials.SecretRef
-		region = pc.Spec.Region
-	case cr.GetProviderReference() != nil:
-		p := &aliv1alpha1.Provider{}
-		if err := c.client.Get(ctx, types.NamespacedName{Name: cr.Spec.ProviderReference.Name}, p); err != nil {
-			return nil, errors.Wrap(err, errGetProvider)
-		}
-		sel = p.Spec.CredentialsSecretRef
-		region = p.Spec.Region
-	default:
-		return nil, errors.New(errNoProvider)
+	if err := c.usage.Track(ctx, mg); err != nil {
+		return nil, errors.Wrap(err, errTrackUsage)
 	}
 
-	if sel == nil {
-		return nil, errors.New(errNoConnectionSecret)
+	if cr.Spec.ProviderConfigReference == nil {
+		return nil, errors.New(errProviderConfigRefNotReady)
```

Comment:
It should be pretty safe to remove this check. In practice `managed.Reconciler` will always make sure this is non-nil before we get here. In the next release of crossplane-runtime (and in master now) we use CRD schema defaulting to make sure it's set, so there are even stronger guarantees that it won't be nil.

Reply:
Removed the check. Ensuring it not to be nil in runtime is great.